synapse/handlers/presence.py

```diff
@@ -1141,6 +1143,33 @@ def _push_presence_remote(self, user, destination, state=None):
             content={"push": [user_state, ], }
         )
 
+    def get_all_presence_updates(self, last_id, current_id, limit):
+        """Return all the presence updates happened since the last id"""
+        # TODO: Respect the limit.
+        # TODO: Find a way to do this without scanning the entire presence map.
+        rows = []
+        for user, cached in self._user_cachemap.items():
+            if last_id < cached.serial and cached.serial <= current_id:
+                rows.append((
+                    cached.serial,
+                    user.to_string(),
+                    encode_canonical_json(cached.state),
+                ))
+
+        for serial, user_ids in self._remote_offline_serials:
+            if serial <= last_id:
+                break
```

Comment:
I think the ``_remote_offline_serials`` are in reverse serial order so we can terminate the loop early here.